src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Options;
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Http;
+using System.Numerics;
+using System.Text;
+using Microsoft.Net.Http.Headers;
+using System.Text.RegularExpressions;
+using System.Globalization;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.HttpLogging
+{
+    internal class W3CLogger : IDisposable
+    {
+        private readonly W3CLoggerProcessor _messageQueue;
+        private readonly IOptionsMonitor<W3CLoggerOptions> _options;
+        // Subtract 8 to account for flags that represent groups (e.g. "RequestHeaders", "All", "None")
+        private readonly int _fieldsLength = Enum.GetValues(typeof(W3CLoggingFields)).Length - 8;
+
+        internal W3CLogger(IOptionsMonitor<W3CLoggerOptions> options)
+        {
+            _options = options;
+            _messageQueue = InitializeMessageQueue(_options);
+        }
+
+        // Virtual for testing
+        internal virtual W3CLoggerProcessor InitializeMessageQueue(IOptionsMonitor<W3CLoggerOptions> options)
+        {
+            return new W3CLoggerProcessor(options);
+        }
+
+        public void Dispose() => _messageQueue.Dispose();
+
+        public void Log(IReadOnlyCollection<KeyValuePair<string, string?>> state)
+        {
+            _messageQueue.EnqueueMessage(Format(state));
+        }
+
+        private LogMessage Format(IEnumerable<KeyValuePair<string, string?>> state)
+        {
+            var elements = new string[_fieldsLength];
+            foreach(var kvp in state)
+            {
+                var val = kvp.Value ?? string.Empty;
+                switch (kvp.Key)
+                {
+                    case nameof(HttpRequest.Method):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Method)] = val.Trim();
```

Comment:
Make a static index that maps Enum values to indices, use that instead of all these Log2's

Reply:
Use length of that for length of array